addons/web/static/src/legacy/js/public/public_root.js

```diff
@@ -82,10 +82,8 @@ export const PublicRoot = publicWidget.RootWidget.extend({
      * @override
      */
     start: function () {
-        var defs = [
-            this._super.apply(this, arguments),
-            this._startWidgets()
-        ];
+        var def = this._super.apply(this, arguments);
+        this._startWidgets();
```

Comment:
@Guillaume-gdi That's not what we want:
- First we have to consider async operations. The public root instantiates the widgets (which is an async operation), it's normal that the start awaits it (the previous code was right).
- This fixes the issue... but only partially, you report the problem: if the public root itself crashes, I still don't have a traceback after your commit, while I should (as in 13.0: if I add a crashing line as the first line of the `start` of the public root, I have a traceback in 13.0).

If you struggle finding the solution, don't hesitate to discuss that with SAD since your found out that one of his commit created the issue (if you are sure that's the cause? you tested before/after that commit to confirm?).

Reply:
@sdegueldre
cc @rdeodoo (I guess we should discuss this :grimacing:)

> tracebacks in dialogs are for runtime errors.

I guess I would prefer a crash during the instantiation of the public root be considered a runtime error then. But could be a matter of taste :shrug: 

> Fixing the issue completely would require making the root.widget module synchronous

Can't do that in stable though unfortunately.

> This current change here simply prevents any public widget crashing from crashing the public root which seems like a good idea to me.

At the cost of simply ignoring the fact it is an async operation, which is not good IMO. I agree the side-effects here could be negligible in master but definitely not in stable (someone requiring the root.widget and do some stuff with it would start doing it before public widgets are instantiated which is not a stable-ok change at all). But even in master, I don't like ignoring async operations: something should await all of them ideally (even if just an exposed mutex). After this change, I have no possibility of placing my potential operations to be done "once the public widgets are instantiated" at the right moment (even making it a dedicated method for the initial start of public widgets that people could `include` does not solve the problem: if I extend that, I can put my operation "after" the public widgets are instantiated, but that moment is still not determined related to all other operations (race conditions...)). By doing that I am actually not sure this cannot create simple bugs, something like this:

Before this commit:
- Load a page
- Public widgets are started, the public root is started, the "edit" button of website is now clickable
- Enter edit, public widgets are first destroyed

After this commit:
- Load a page
- The public root is started, the "edit" button of website is now clickable.
- Public widgets are *being* started
- Enter edit, public widgets are destroyed before they were even finished starting, potential crash/bug. (I don't know if that's actually the case after this commit or even if it was not already a potential problem before this commit since the `_super` call is done "in parallel" (I'll check when I get the time) but I am just illustrating that not awaiting async operations at all are more often problems than awaiting them even "randomly" IMO).

> Previously, an error occurring during the definition of an async module would not mark the module as having failed to load

I wanted to check that module loading info on crash, I indeed have the error log but nothing in the "debug" module log part.
Is that normal @sdegueldre ? I think it was the case before :shrug: 

> If you wanted it to have the the same behaviour as previously in the happy path and not crash if any public widget crashes you could also just wrap the public widget promises in a Promise.allSettled().

Seemed like a good idea but I tried and ended not having a crash in the console / module loading and not having a traceback, maybe I did something wrong :shrug: 

As a stable fix, I don't have a better idea than this right now:
```javascript
        var defs = [
            this._super.apply(this, arguments),
            this._startWidgets().catch(err => {
                setTimeout(() => {
                    throw err;
                });
            })
        ];
```
Which is obviously ugly but has the merit of:
- Not changing async flows
- Errors in public widgets trigger a traceback

... but maybe it's not worth a fix in stable at all, I am not sure: we could be "reviving" bugs in visitors pages which were silent before... but can't imagine something making the public root crash that was silently not inducing bigger problems than a traceback. #overthinking

Reply:
Right, I can see that there is definitely merit to having access to "the public widgets are started" promise, and I'm also not opposed to tying that to the start of the public root itself, but yeah, so long as this widget's start promise is returned by a module it means all errors that happen while it or the public widgets are starting happen during module loading which means dependent modules never get loaded which is less than ideal. The .catch into setTimeout to throw the error in a different call stack can work, thinking about it, maybe we can do that in the module system instead. We definitely don't want to stop processing modules as soon as a module crashes so throwing an error there which was Guillaume's original proposal isn't what we want but throwing it in a different call stack should work, I'll play around with this idea.

Reply:
Hello @sdegueldre, did you have the time to play around with that idea? :)

Reply:
@qsm-odoo I haven't and I don't foresee having time to do it in the next few weeks. I've made a note for it with a link to this PR so we can get to it when we're in less of a rush. If the issue is blocking, let me know and I'll bump it up in priority.

Reply:
I am going through our PRs to sort it a bit, was there anything new following this discussion?
I didn't read it but just the last message, any update about this so we can see how we can move forward here?

Reply:
Completely lost track of this. I've opened a PR with the approach I was entertaining: https://github.com/odoo/odoo/pull/115052

This restores the error dialog when a public widget errors while starting and doesn't crash the module system. Can you and/or qsm check that this solves the issue as you were expecting? I might still need to discuss the details with the rest of the team but would be good to know if this approach works for you before doing that.

Reply:
@sdegueldre Tested, seems to work fine, thanks :+1: :crossed_fingers: